### PR TITLE
Fix release post contributors

### PR DESCRIPTION
**Pull request by lsinger**
### Submission Review Guidelines:

- I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
- I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
- I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/). 
- Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

### Changes proposed in this Pull Request:

When releasing 7.6 I noticed that the `release-post` tool exits with an error when trying to generate the release post, notably at the part where it's collecting the contributor list. This PR fixes this and adds a few other tweaks that made the generated post more useful, such as removing whitespace to enable proper copy/paste in the block editor. 

### How to test the changes in this Pull Request:

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. check out `trunk`
2. `cd` into `tools/release-post`
3. run `pnpm run release-post -- "7.6.0" "7.5.1" --outputOnly`
4. notice how this fails when trying to collect the contributor list
5. check out this branch and repeat the process
6. notice how the release post is now being generated

### CODE REVIEW

**Review by jonathansadowski — DISMISSED** [review]
Thanks for cleaning this up. This works well. Just noting that I had to test with 7.7.0/7.6.0 to reproduce the issue on trunk, because the previous command worked with 7.6.0/7.5.1 now that 7.6.0 is tagged.

Would you mind applying the same fix you made here to `tools/release-posts/commands/release-post/release-post-contributors.ts`? It suffers from the same issue you fixed in the main command.

**Comment by lsinger** [comment]
Thanks @jonathansadowski, made the requested changes. I thought about moving the version → branch conversion to its own module, but things are a bit entangled as the release-post command also clones the repo as part of its branch-finding process. For now I simply cloned parts of the code, which we can then iterate on later. 

The commands to use for the review are now: 

`pnpm run release-post -- "7.7.0" "7.6.0" --outputOnly`

and

`pnpm run release-post contributors "7.7.0" "7.6.0"`

**Review by jonathansadowski — APPROVED** [review]
Thanks for updating the other command... looks good to me, nice work!

File: tools/release-posts/commands/release-post/release-post-contributors.ts [hunk 1]
```diff
@@ -28,9 +29,37 @@ const program = new Command()
...
+			if ( ! previousParsed ) {
+				throw new Error( 'Unable to parse previous version' );
+			}
+			previousBranch = `release/${ previousParsed.major }.${ previousParsed.minor }`;
+		} catch ( error: unknown ) {
+			throw new Error(
+				`Unable to find '${ previousBranch }'. Branch for previous version must exist.`
+			);
+		}
+
```
Comment (samueljseay) [inline tools/release-posts/commands/release-post/release-post-contributors.ts:59]: Neither of these should be try/catch, it increases confusion in the code paths here (much harder to quickly tell what's happening since you keep throwing and catching as a way to handle different code paths). You don't need try/catch unless you're calling APIs that by behaviour will throw exceptions in some states.

Instead you should just check these conditions and when you want to crash the CLI call `Logger.error` which will log and exit.
```suggestion

	if (  ! currentParsed ) {
		Logger.notice( 'Unable to parse current version' );
		currentBranch = 'trunk';
	} else {
	    currentBranch = `release/${ currentParsed.major }.${ currentParsed.minor }`;
	}
	
	if ( ! previousParsed ) {
	  // Logger error does process.exit and logs the message
	  Logger.error(`Unable to find '${ previousBranch }'. Branch for previous version must exist.`)
	} else {
	  previousBranch = `release/${ previousParsed.major }.${ previousParsed.minor }`;
	}


```

File: tools/release-posts/templates/release.ejs [hunk 2]
```diff
@@ -11,25 +11,13 @@
...
-      <li>
-        <a target="_blank" href="https://github.com/<%= org %>/<%= repo %>/compare/<%= baseRef %>...<%= headRef %>">
-          <%= totalCommits %> commits from <%= packageContributors.length %> contributors
-        </a> 
-        in <%= packageName %>.
-      </li>
-    <% } %>
-  <% } %>
-</ul>
+<ul><% for ( const packageName in contributors ) { %><% const {org, repo, contributors: packageContributors, totalCommits, baseRef, headRef} = contributors[packageName] %><% if ( packageContributors ) { %><li><a href="https://github.com/<%= org %>/<%= repo %>/compare/<%= baseRef %>...<%= headRef %>"><%= totalCommits %> commits from <%= packageContributors.length %> contributors</a> in <%= packageName %>.</li><% } %><% } %></ul>
```
Comment (samueljseay) [inline tools/release-posts/templates/release.ejs:14]: This formatting is not good, we need the ejs to be easy to edit and maintain and read, I use this extension for ejs support in VSCode https://marketplace.visualstudio.com/items?itemName=DigitalBrainstem.javascript-ejs-support

File: tools/release-posts/templates/release.ejs [hunk 3]
```diff
@@ -44,7 +32,7 @@
 <!-- wp:paragraph -->
 <p>
   These are just some of the changes that are included in WooCommerce <%= displayVersion %>. 
-  You can find the complete changelog for this release in the <a href="https://github.com/woocommerce/woocommerce/blob/release/<%= displayVersion %>/changelog.txt" target="_blank">changelog.txt file</a>.
+  You can find the complete changelog for this release in the <a href="https://github.com/woocommerce/woocommerce/blob/<%= currentBranch %>/plugins/woocommerce/readme.txt">readme.txt file</a>.
```
Comment (samueljseay) [inline tools/release-posts/templates/release.ejs:35]: Why did you remove target blank? Its quite convenient imo to have links open new tab especially when reading and jumping between documentation.

**Review by samueljseay — CHANGES_REQUESTED** [review]
I think the try/catch code should be replaced with regular conditionals

File: tools/release-posts/templates/release.ejs (hunk 2 above)
Comment (lsinger) [reply tools/release-posts/templates/release.ejs:14]: @samueljseay I deliberately formatted things this way because with the old, more readable formatting, pasting into Gutenberg didn't work for me. I didn't like all those new lines in all the places. Maybe there's a different way to change the output here so that it works for copy and paste?

File: tools/release-posts/templates/release.ejs (hunk 3 above)
Comment (lsinger) [reply tools/release-posts/templates/release.ejs:35]: @samueljseay generally my opinion is that the user should decide on their end how and where they'd like a link to open. E.g., I usually use Command-click if I want things to load in a background tab. I had assumed the basic philosophical stance behind it ("let the client decide this") to be fairly uncontroversial, so I did not comment on this initially. Happy to reconsider, but fundamentally I don't think we can and should be deciding which links all readers will want to open in a new window.

File: tools/release-posts/commands/release-post/release-post-contributors.ts (hunk 1 above)
Comment (lsinger) [reply tools/release-posts/commands/release-post/release-post-contributors.ts:59]: @samueljseay looking at it now I do agree that the exception strategy here seems a bit odd. I took it from `release-post-release.ts`. I'd like to accept your suggestion and then rework things in `release-post-release.ts` in a similar manner as well, but before doing that I'd like to know whether @jonathansadowski maybe had a specific rationale for writing it this way?

File: tools/release-posts/templates/release.ejs (hunk 2 above)
Comment (samueljseay) [reply tools/release-posts/templates/release.ejs:14]: @lsinger damn, thats frustrating :/ that seems like a sort of big issue, because ideally we would have nice readable/maintainable templates (that was why I moved away from the string concatenation mess we had in the past)

I think we have 2 things available in EJS we should try before making templates harder to read.

1. pass the option `rmWhitespace` to the render call (see `Options` [here](https://ejs.co/)) - you can pass it to `renderFile` [here](https://github.com/woocommerce/woocommerce/blob/5ba15f84cd39a36374d0fb0c669abd1a6d5cef41/tools/release-posts/lib/render-template.ts#L17).

2. There are also whitespace slurping tags: `<%_` and `_%>`(they strip whitespace before and after them not inside) but probably just doing rmWhitespace makes more sense.

File: tools/release-posts/templates/release.ejs (hunk 3 above)
Comment (samueljseay) [reply tools/release-posts/templates/release.ejs:35]: @lsinger no probs, just a philosophical difference and I'm fine with that. I think the opinions seem to differ but for reference material when reading an article I think it's handy since you don't want to leave the reference material while reading. That said I see super mixed opinions on it, so happy to remove 😄

**Comment by lsinger** [comment]
Closing as this isn't an issue anymore.
